### PR TITLE
Adds composite tracer factory to blockchain processor.

### PR DESCRIPTION
**Pull request by MicahZoltu**
This makes it so a plugin can add a block tracer to all new blocks that get processed.  I added it directly to the BlockchainProcessor for two reasons:
1. Adding a new constructor parameter to BlockchainProcessor requires a *ton* wiring.  I can do this if it is desired, but I wanted to start with the simplest solution and get feedback before putting in all of that legwork.
2. `CompositeBlockTracerFactory` is a very thin class (just a reference and a bool) and isn't interesting enough IMO to be considered a generalized service.  In fact, it *may* make sense to just integrate it directly into BlockchainProcessor.

`EndBlockTrace` was added to `IBlockTracer` so that plugins can correctly aggregate transaction traces by block.  The block is passed as a parameter to this callback just to make it easier on the plugins to correlate start and end of the block trace.

Usage:
```cs
api.BlockchainProcessor.BlockTracerFactory.AddChildFactory(myBlockTracer);
```
This will cause `myBlockTracer` to be attached to every new block that is processed by the primary block processing pipeline.  In the case where a plugin wants to trace every block (e.g., for building a block explorer), this avoids having to unnecessarily execute the block multiple times (once without tracers, and then again with tracers).

No tests included here, but I can add them if this change otherwise seems reasonable to the maintainers.

### CODE REVIEW

**Review by tkstanczak — COMMENTED** [review]
@MarekM25 do you think you could suggest testing and architectural review here. I think this makes sense but maybe better to have a NyullFactory by default and allow to pass another? also what would be the best way to cover it with tests?
is the EndTrace risky in any way inside TraceBlock in Parity tracer calls? in Geth tracer calls?

**Comment by MarekM25** [comment]
Hi @MicahZoltu 
I wonder if we need a parameter for EndBlockTrace. Could you just save the block in your plugin after start instead of passing it? Please note, if you want to add EndBlockTracer you should also add it in several places. For example here: https://github.com/NethermindEth/nethermind/blob/master/src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs#L56

I will consider change:
`api.BlockchainProcessor.BlockTracerFactory.AddChildFactory(myBlockTracer);`
to
`api.BlockchainProcessor.AddTracer(myBlockTracer);`

You can create an interface IBlockTracerExtendable (naming is just a suggestion) and extend the IBlockchainProcessor interface with it. In this interface you could have methods:
```
void AddTracer(IBlockTracer tracer);
void RemoveTracer(IBlockTracer tracer); //consider using tracer ID that would be returned by AddTracer and used by RemoveTracer
```
In the blockchain processor you could have a private IBlockTracerProvider (or IBlockTracerFactory) field with methods:
```
IBlockTracer GetTracer();
void AddTracer(IBlockTracer tracer);
void RemoveTracer(IBlockTracer tracer); 
```
Implementation of IBlockTracerProvider could return NullBlockTracer by default and if we add at least one Tracer we could start return CompositeBlockTracer

Of course, everything I wrote is just a proposition.

**Comment by MicahZoltu** [comment]
> I will consider change:
api.BlockchainProcessor.BlockTracerFactory.AddChildFactory(myBlockTracer);
to
api.BlockchainProcessor.AddTracer(myBlockTracer);

Wouldn't this be an incorrect behavior change?  I *believe* (please correct me if I'm wrong) that we need a new tracer for every block, which means we need to be adding a tracer factory, not a tracer to the `BlockchainProcessor`?

If you just mean that we should get rid of the layer of abstraction and put `AddTracerFactory` directly on `BlockchainProcessor` that is reasonable and I can do that.  😄

**Comment by LukaszRozmej** [comment]
> > I will consider change:
> > api.BlockchainProcessor.BlockTracerFactory.AddChildFactory(myBlockTracer);
> > to
> > api.BlockchainProcessor.AddTracer(myBlockTracer);
> 
> Wouldn't this be an incorrect behavior change? I _believe_ (please correct me if I'm wrong) that we need a new tracer for every block, which means we need to be adding a tracer factory, not a tracer to the `BlockchainProcessor`?
> 
> If you just mean that we should get rid of the layer of abstraction and put `AddTracerFactory` directly on `BlockchainProcessor` that is reasonable and I can do that. 😄

Yes we need to instantiate new tracers. We can have AddTracerFactory on BlockchainProcessor, please keep interface segregation  principle in mind.

**Comment by MarekM25** [comment]
> Wouldn't this be an incorrect behavior change? I _believe_ (please correct me if I'm wrong) that we need a new tracer for every block, which means we need to be adding a tracer factory, not a tracer to the `BlockchainProcessor`?
> 
> If you just mean that we should get rid of the layer of abstraction and put `AddTracerFactory` directly on `BlockchainProcessor` that is reasonable and I can do that. 😄

Got it. You're right. We need a new tracer for every block. Hmmm IMO, it will be more elegant to have API like:
`api.BlockchainProcessor.AddTracerFactory(myBlockTracerFactory);`

But please do not add this method to IBlockchainProcessor and you should create a separate interface for it.

**Comment by MicahZoltu** [comment]
> I wonder if we need a parameter for EndBlockTrace. Could you just save the block in your plugin after start instead of passing it?

The idea was to use the block essentially as a correlation identifier for the start and end calls.  However, in hindsight I'm struggling to come up with a situation where you would need that correlation because `StartNewBlockTrace` should be called exactly once for a given tracer, and `EndBlockTrace` should be called exactly once.  I have removed the `Block` parameter.

> Hmmm IMO, it will be more elegant to have API like:
> `api.BlockchainProcessor.AddTracerFactory(myBlockTracerFactory);`
> 
> But please do not add this method to IBlockchainProcessor. You should create a separate interface for it.

`BlockchainProcessor` on `NethermindApi` is an `IBlockchainProcessor`.  If I don't put the `AddTracerFactory(...)` onto `IBlockchainProcessor`, then we won't be able to do `api.BlockchainProcessor.AddTracerFactory(...)`.  I could add the new interface onto `NethermindApi` and just set it to the blockchain processor.

Out of curiosity, what is the reasoning for keeping the `AddTracerFactory` off of the `IBlockchainProcessor`?  It *feels* like it is a method that is tightly coupled with a BlockchainProcessor, as you are telling the blockchain processor "when you process a block, please run with tracers from this factory".  I suspect I'm misunderstanding some piece of the architecture though and maybe this doesn't make sense?

**Comment by MarekM25** [comment]
> > I wonder if we need a parameter for EndBlockTrace. Could you just save the block in your plugin after start instead of passing it?
> 
> The idea was to use the block essentially as a correlation identifier for the start and end calls. However, in hindsight I'm struggling to come up with a situation where you would need that correlation because `StartNewBlockTrace` should be called exactly once for a given tracer, and `EndBlockTrace` should be called exactly once. I have removed the `Block` parameter.
> 
> > Hmmm IMO, it will be more elegant to have API like:
> > `api.BlockchainProcessor.AddTracerFactory(myBlockTracerFactory);`
> > But please do not add this method to IBlockchainProcessor. You should create a separate interface for it.
> 
> `BlockchainProcessor` on `NethermindApi` is an `IBlockchainProcessor`. If I don't put the `AddTracerFactory(...)` onto `IBlockchainProcessor`, then we won't be able to do `api.BlockchainProcessor.AddTracerFactory(...)`. I could add the new interface onto `NethermindApi` and just set it to the blockchain processor.
> 
> Out of curiosity, what is the reasoning for keeping the `AddTracerFactory` off of the `IBlockchainProcessor`? It _feels_ like it is a method that is tightly coupled with a BlockchainProcessor, as you are telling the blockchain processor "when you process a block, please run with tracers from this factory". I suspect I'm misunderstanding some piece of the architecture though and maybe this doesn't make sense?

What was on my mind is to create a new interface (for example IBlockTracerExtendable) with methods:
```
void AddTracerFactory(IBlockTracerFactory tracer);
void RemoveTracerFactory(IBlockTracerFactory tracer);
```

and then IBlockchainProcessor : IBlockTracerExtendable

**Comment by MicahZoltu** [comment]
Sorry for dragging my feet on this.  It isn't because of any feedback you all gave, just some other things came up that caused this work to get re-prioritized lower.  😢  I still intend to return to this once time permits, unless someone beats me to it.

File: src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+namespace Nethermind.Evm.Tracing
+{
+    public class CompositeBlockTracer : IBlockTracer
+    {
+        private readonly ImmutableArray<IBlockTracer> _childTracers;
+        public bool IsTracingRewards { get; private set; }
+
+        public CompositeBlockTracer(IEnumerable<IBlockTracer> childTracers)
+        {
+            _childTracers = ImmutableArray<IBlockTracer>.Empty.AddRange(childTracers);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs:17]: usage of Immutable array seems off

File: src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+using Nethermind.Int256;
+
+namespace Nethermind.Evm.Tracing
+{
+    public class CompositeBlockTracer : IBlockTracer
+    {
+        private readonly ImmutableArray<IBlockTracer> _childTracers;
+        public bool IsTracingRewards { get; private set; }
+
+        public CompositeBlockTracer(IEnumerable<IBlockTracer> childTracers)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs:15]: move to array everywhere, index access is faster than enumerable

File: src/Nethermind/Nethermind.Blockchain/Processing/IBlockchainProcessor.cs [hunk 3]
```diff
@@ -23,6 +23,8 @@ namespace Nethermind.Blockchain.Processing
 {
     public interface IBlockchainProcessor : IDisposable
     {
+        public CompositeBlockTracerFactory BlockTracerFactory { get; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Processing/IBlockchainProcessor.cs:26]: remove public from interface

**Review by LukaszRozmej — COMMENTED** [review]
@kristofgazso can you add tests and fix few issues and then merge?

File: src/Nethermind/Nethermind.Blockchain/Processing/OneTimeProcessor.cs [hunk 4]
```diff
@@ -24,6 +24,8 @@ namespace Nethermind.Blockchain.Processing
 {
     public class OneTimeChainProcessor : IBlockchainProcessor
     {
+        public CompositeBlockTracerFactory BlockTracerFactory { get; } = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Processing/OneTimeProcessor.cs:27]: This looks unused, shouldn't it point to inner processor in any way?

**Review by LukaszRozmej — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]

**Comment by MicahZoltu** [comment]
Does this need anything from me to proceed since I'm the original author, or can it continue to progress without me?  If it matters, you can consider everything I wrote optimally permissively licensed (CC0, MIT, Apache, WTFPL, Unlicense, or whatever makes it so you can just ignore licensing for this PR).

File: src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/Tracing/CompositeBlockTracer.cs]: this is not resolved

**Comment by MicahZoltu** [comment]
Thanks everyone!

**Pull request merged**